### ISSUE TITLE
Windows - creating bitmap fonts programmatically

### ISSUE THREAD

**Issue by Mikhail22** [fontforge/fontforge#3238]
Hello all.

I am new to Fontforge and it's my first post here. 
(I was about to ask on nabble forum but it seems it's in archive mode)
So I think it is ok to post a question here, and it is merely a general font 
creation question, but probably you can guide me. 

I want to create a bitmap-only font on Windows, so it can be registerd in platform.
I already have bitmaps, and everything for that.

What I have found out:
1) There is a (legacy?) FON/FNT format. I have some of those on my PC, 
but they seem to show-up only in 1-bit, though MS docs say:

DFF_1COLOR means 8 pixels per byte
DFF_16COLOR means 2 pixels per byte
DFF_256COLOR means 1 pixel per byte
DFF_RGBCOLOR means RGBquads

So it might support graymaps also, I am not sure. I have read the docs, 
but now I am thinking -- even if it supports graymaps, then it still 
does not support kerning and glyph overlapping (correct me if I am wrong), 
So this option is out, because there is no chance to avoid cutted pixel sides 
or setting up kerning. But maybe there is some newer reincarnation of that format?

2)  There is a TTF 'bitmap strikes' - this info I have found on the FontForge main page.
So maybe this is what I need... I hope so, and if so, then my question:
Is it possible to generate such Windows-compatible font from my bitmaps (8-bit maps)?

*If yes*, still I want to note the requirements (so as not to end up in misunderstanding):

 - must be possible to compile it *programmatically*, because otherwise I will spend a good
part of my life with only one font.
 - bitmaps (i.e. alpha-maps) must be 1:1 preserved on the display, i.e. without any distorion, scaling, etc.
 - glyph overalapping must be supported of course
 - possible to define kerning pairs in pixels (or other values)

I have all bitmap data, kernings, sizes, etc for that and 
I can adopt it to suit the tool input format if needed.

So please guide me in the right direction.

Thanks in advance,
Mikhail

**Comment by ilyaza** [comment]
- There may be another group for *“User”* questions.
 - What makes you think you need ***bitmap*** fonts?
 - You may want to have a peek at http://ilyaz.org/software/fonts

**Comment by Mikhail22** [comment]
@ilyaza I create bitmap fonts optimized for screen and want to try compile them for platform-wide usage. Regarding the link you provide - I am not sure what it has to do with the question?

**Comment by ilyaza** [comment]
“bitmap fonts optimized for screen” — what may it mean?!

If this turn of speach is due to you believing that the best fonts for screen are bitmap fonts — you are way wrong.  AFAIK, with contemporary rasterizers scalable fonts should provide much better rendering.

Which leads to the link I provided.  The build tools are in the “non-mono” archive.

**Comment by lemzwerg** [comment]
@ilyaza, your bold statement isn't true in general, especially if you don't have a high-DPI screen.  In particular, CJK fonts at small sizes can only be displayed in a legible way on such displays if you use handcrafted bitmaps (or graymaps, as high-end companies like Morisawa provide for some customers).

However, I agree that this discussion doesn't belong into the bug tracker of fontforge; the issue should rather be discussed on the [user mailing list](https://sourceforge.net/projects/fontforge/lists/fontforge-users).

**Comment by ilyaza** [comment]
@lemzwerg: I would love to hear more about such fonts — and how they may be rendered better by hand in B/W than automatically in color.

In fact, the reference which I did is almost exactly about this: I take a **very low resolution** raster font (with CJK too!), and convert it to outline — and AFAICS the rendering on Windows improves a lot after this.  (This assumes that my algorithms recognize correctly what are the strokes in a font — but recently they are becoming better and better in this.)

**Comment by lemzwerg** [comment]
@ilyaza, what you do is very special and not suited for the general case – there are good reasons that even today CJK fonts contain bitmap strikes...

At small sizes, you have to change the *topology* of CJK glyphs to make them readable.  If you have, say, twelve horizontal strokes, you get problems if rendering at 14x14px.   Have a look at the [ideohint](https://github.com/caryll/ideohint) project that successfully tries to simplify glyphs in an automated way, but manually created bitmaps are of course still better.

**Comment by Mikhail22** [comment]
@lemzwerg thanks for a link, I have posted this question there right now.
@ilyaza So handcrafted bitmap image in general will be always better than something rasterized automatically, especially for details on lower PPI displays it is critical. I don't know what to add to this, and frankly, it seems pretty self-evident to me. Anyway my question is not even about the readability or quality. I am searching about technical possibilities for now.

**Comment by ilyaza** [comment]
@Mikhail22 “handcrafted bitmap image in general will be always better than something rasterized”.

BS.  If you think otherwise, beat this:  On my system,
`    m w m w m M W ɯ ʍ m w Ɯ ш щ Ш Щ ⧢`
shows easily when I set the character bbox on my console to 6×3 pixels.  Overwhelming majority of the shown characters can be distinguished pairwise without much stress.  (But if I use `Magnifying glass` application, it shows a complete mess.)  IIRC, my setup is in the default rendering settings.

Do this with a bitmap font.

@lemzwerg Sorry for being so dense, but I have no clue how a change in topology when changing pixel size is related to difference between bitmap/outline fonts.  Both technologies support such changes without any problem.  So given that at low DPI outline fonts have a major chance to be shown clearer than bitmap/graymap fonts, I do not see how your argument may work.

**Comment by Mikhail22** [comment]
@ilyaza You seem to continue to mis-interpret the whole topic beeng discussed. One more time: Bitmap fonts, bitmap strikes, etc. all refer to *non-scalable* fonts, unless stated the opposite. And I have explicitly stated in the question that I do 1:1 non-sclable, full per-pixel alpha fonts. And yes, this is the way to achieve sharp glyphs, and manual perfectizing of whole look, contrast, stroke transitions, etc. But of course there are a lot of other approaches, but thats another topic.

**Comment by Mikhail22** [comment]
@ilyaza In other words, for a given letter pixel size, a rasterized path cannot compete with a handcrafted bitmap, not even in theory. This is true for higher PPI display aslo, but on a really high PPI (>300) display the difference will not be so noticable, but it is still there.

**Comment by ilyaza** [comment]
@Mikhail22 You continue to mis-interpret what I said.  It is a very simple thing.

 - You have a theory.  Fine.  However, AFAIK, this theory is completely bogus.  
    It may have been true in 1990 (ATM just being born); nowadays, with ⅓-pixel positioning of strokes, this is way out of question.
 - There are two technologies: raster fonts, and outline fonts.  With rasterizers of today, outline fonts provide much better whole look, contrast, stroke transitions, etc than bitmap fonts (when one compares them **at the same size**).
 - This advantage of outline fonts is especially pronounced with low DPI displays.
 - AFAIK, there is very little chance that a manually created bitmap font would match a “performance” of auto-rasterized outline font (at the same size).  Same holds for manually created graymaps.
 - This is completely orthogonal to outline fonts being (in addition to the advantages stated above) scalable.

If you do not believe in this, I proposed a very simple experiment.  I repeat it here: see how 
    `w m M W ɯ ʍ Ɯ ш щ Ш Щ ⧢`
looks with `DejaVu Sans Mono` when rendered into a character box of 6×3 pixels.  Try to reproduce this with a bitmap font.  Just for your convenience, here is the template for you to edit (with `i` and `j` drawn as an example):
```
    -#-  --#  ---  ---  ---  ---  ---  ---  ---  ---  ---
    ---  ---  ---  ---  ---  ---  ---  ---  ---  ---  ---
    -#-  --#  ---  ---  ---  ---  ---  ---  ---  ---  ---
    -#-  --#  ---  ---  ---  ---  ---  ---  ---  ---  ---
    -#-  #-#  ---  ---  ---  ---  ---  ---  ---  ---  ---
    ---  -#-  ---  ---  ---  ---  ---  ---  ---  ---  ---
```
The aim: all the chars should be easily recognizable, and most should be different from each other.

―――――――――――――――――――――――――

**P.S.** Above, what I say is “AFAIK”.  If I’m wrong, I would gladly follow pointers able to convince me otherwise.  But keep in mind that 10 years ago I thought the same way as Mikhail (also basing this on “theoretical considerations”) — it is practice which taught me otherwise.

**Comment by Mikhail22** [comment]
@ilyaza Dude, that seems to be quite surrealistic talk. What pattern do you want me to reproduce and why? Which bitdepth you mind and what composer/blending mode? There is still one option, that maybe you still believe that I was talking about 1 bpp fonts or something? But even that is quite strange because I explicitly stated right in my question that I work with full 8-bit alpha channel, and in general, I have other approaches in my custom composer with multi-layering, etc. I wonder, if that is not the case with your misunderstanding (which I would see as complete ignorance because everything should be clear from the question). And if it's not, then what the hell are you talking about here?

**Comment by ilyaza** [comment]
I’m talking about ⅓-pixel positioning.

**Comment by ilyaza** [comment]
Let me repeat it the ***THIRD*** time: the discussed effect is:
     `Outline fonts can show the string mentioned above using 6×3 character cells` (“relatively crisp”).
You claimed: “bitmap image in general will be always better”.  Prove it.  Show how to render these 12 characters each into a 6×3 graymap so that they are as easy identifiable as with `DejaVu Sans Mono`.

————————————————————————————

BTW, I have no clue what your question about bitdepth means.  Last time I’ve read about effects of bitdepth on anti-aliasing was about 20 years ago.  At the time, it was thought (IIRC) that anything above 4 bits is an overkill.  Do you know otherwise?

**Comment by Mikhail22** [comment]
I don't need to prove anything, that is you who argues about something that is not directly related to applications that I do. I am focused on applications where cognitive qualities are critical, i.e. where a user is reading a lot and for a long time. something 'relatively crisp' is not telling anything. If you ask, in my case there is zero practical usage of monospaced 6x3 font. If you want me to show how to implement something particular, then make a 1:1 screenshot with clear explanation of criterias and requirements, then I may be interested.

**Comment by Mikhail22** [comment]
So you claim that *your* (as I am not still sure if it is a part of some platform) method of 1/3 placement gives good (and good can refer to many different aspects of quality) results in such text quality, so that I, having ability to manipulate source data on any level, cannot achieve and beat? If that is your claim, then you may be opening some new page in history. And, even whithout seeing your rendered text examples, I'll be brave enough to say that you are flat false. And if your claim is just - "some rasterizers gives good quality", then it is just handwaving and does not tell *anything*. So I'd say it is you who should show some rendered text examples here and not me, for frequently used sizes, i.e. with a font with x-height of 5 or 6 px (assuming 96 PPI display).

**Comment by Mikhail22** [comment]
In a wish to make your further arguing a bit more mature I will give you a hint - what makes you think that I, dealing only with bitmap data, cannot implement 1/3 placement, 1/4 placement ... 1/10 placement?

**Comment by JoesCat** [comment]
Hi @Mikhail22,
Reading your original posting:

In terms of nabble, you may want to use the sourceforge mailing list.
My past experience with downstream mailinglists are they sometimes are out of sync, and what gets posted downstream does not necessarily make it's way back upstream.
Look here: https://sourceforge.net/p/fontforge/mailman/fontforge-users/
You can also search here for additional info, sometimes a question has been asked at some point in the past, therefore you might find info this way.

Additional information can be found by reading the howto:
https://github.com/fontforge/designwithfontforge.com
Many developers have worked to create this, but I'm sure more can be added, and if you see info you learned and believe can also be added to help future FontForge users, you are encouraged to add your 2 cents.

I note you have already posted/asked on the user mailinglist, but I'm mentioned this here for others following this thread who are not aware of it.

You are new to FontForge, as old as it appears, FontForge still continues to be a work in progress. A lot has improved, and there is yet room for improvement.

Based on your OP, I suspect you are looking at drag and drop of individual characters to make up a font, or use of native or python scripting to do this programatically. If you need a template, you can check the tests subdirectory as a starting point for a/your script, and the doc/html directories have a jumble of info that can help you with the native and python commands.
https://github.com/fontforge/fontforge/tree/master/doc/html
https://github.com/fontforge/fontforge/blob/master/doc/html/scripting.html
https://github.com/fontforge/fontforge/blob/master/doc/html/python.html

Best place to ask questions is on the user list as the number of eyes watching the user list should outnumber those watching these issues (mainly meant for tracking program bugs, not how to use fontforge).

Hi @ilyaza,
There are times where we have to accept that advice can be good, is noted, but does not always apply. As an example, I recall working at one company where the IT department insisted and demanded acceptance and use of certain programs under threat of getting cut-of network-wise and abandonment of computer backups. Unfortunately for the IT department, the processes that were being used were basically the meat-n-potatoes bringing in the cash for the company, and the IT department had to concede and make allowances and exceptions for these non-standard processes since they had no solutions for these processes.
In summary, what you recommend may be good/great, has been heard and noted, but there can/may be unseen/unknown exceptions for doing what @Mikhail22 asked....my guess is probably a dedicated monitor with a low pixel count for X & Y.

Reading the following 2/3 of this thread, I think we all could benefit without poking the bear any further.
Happy Holidays to everyone :-)

**Comment by ilyaza** [comment]
Happy new year!

For people who ***indeed*** missed all the font rasterizing developments of 3rd millennium, see [these photos](https://en.wikipedia.org/wiki/Subpixel_rendering#Examples).

And: **No**, one cannot illustrate this by doing screenshots on a computer.

**Comment by jtanx** [comment]
Ok, I don't think this is productive to the issue at hand, please move on @ilyaza

**Comment by RicoElectrico** [comment]
Oh God. I just came here googling "fontforge rasterizing a font" to see if it'd be any good to turn open source fonts into bitmaps with a little bit of manual help (to use it in Arduino projects like  [u8g2](https://github.com/olikraus/u8g2) ).

I regret wasting 2 minutes of my life on autistic rants in this thread.

**Comment by ctrlcctrlv** [comment]
@RicoElectrico I regret wasting even five seconds reading your inane comment.

**Comment by ctrlcctrlv** [comment]
(Closing as questions belong on the ML and this thread, even years later, remains troll bait. Anyway, the correct thing to do here is to make a BDF font. FreeType can help you get somewhat acceptable rasters out of outline fonts with hinting. FontForge can convert BDF to a TrueType font with bitmap strikes if that's what you need.)
